### PR TITLE
Rework SnakeCase/KebabCase naming policies to closer match Json.NET's

### LINKED ISSUE #77309: Validate that the new snake/kebab case naming policies match the Json.NET implementation.
We should probably add a few tests validating that the output matches that of the Json.NET implementation, particularly when it comes to nontrivial inputs. We have time until .NET 8 ships, so we can always introduce breaking changes if needed.

_Originally posted by @eiriktsarpalis in https://github.com/dotnet/runtime/pull/69613#discussion_r1001603922_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This PR:

1. Updates the JsonSeparatorNamingPolicy so that non-alphanumeric characters are not being truncated, minimizing the probability of naming collisions.
2. Extends unit test coverage for the component.

After a lot of discussion we concluded that:

1. The policies should keep the fixes for https://github.com/JamesNK/Newtonsoft.Json/issues/1956, resulting in some deviation from Json.NET semantics. We need to document these where appropriate.
2. We will not be adding support for surrogate pairs. Surrogate pairs are handled gracefully by the current implementation, however uppercase surrogate pairs are not recognized as such. This matches the implementation of `JsonNamingPolicy.CamelCase`. We should consider adding surrogate pair support in the future, although that would require addressing https://github.com/dotnet/runtime/issues/52947.

Fix https://github.com/dotnet/runtime/issues/77309.

### CODE REVIEW

**Comment by stephentoub** [comment]
I just left comments on https://github.com/dotnet/runtime/pull/90248. What's the difference between this one and that?

**Comment by YohDeadfall** [comment]
I like it much more than the previous attempt, and it fits into the survey results which isn't representative at the moment. There are two votes on Mastodon for keeping punctuation, but another one from Twitter is for trimming. With the first question things are much better and there are 86% of votes for partial compatibility with JSON.NET on Mastodon and 50% on Twitter (where just two people voted).

**Comment by YohDeadfall** [comment]
> What's the difference between this one and that?

The difference is in how punctuation is treated: https://github.com/dotnet/runtime/pull/90248#discussion_r1288775695. This pull request behaves more naturally.

File: src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs [hunk 1]
```diff
@@ -21,149 +28,154 @@ public sealed override string ConvertName(string name)
...
+                            WriteChar(separator, ref destination);
+                        }
 
-            for (int index = 0; index < chars.Length; index++)
-            {
-                char current = chars[index];
-                UnicodeCategory currentCategoryUnicode = char.GetUnicodeCategory(current);
+                        if (!lowercase)
+                        {
+                            current = char.ToUpperInvariant(current);
```
Comment (tarekgh) [inline src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs:105]: >ToUpperInvariant(current); [](http://example.com/codeflow?start=43&length=26)

do we need to upper case when having DecimalDigitNumber?

File: src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs [hunk 2]
```diff
@@ -21,149 +28,154 @@ public sealed override string ConvertName(string name)
...
-                    currentCategoryUnicode <= UnicodeCategory.OtherPunctuation)
-                {
-                    WriteWord(chars.Slice(first, index - first), ref result);
+                        WriteChar(current, ref destination);
+                        state = SeparatorState.LowercaseLetterOrDigit;
+                        break;
 
-                    previousCategory = CharCategory.Boundary;
-                    first = index + 1;
+                    case UnicodeCategory.SpaceSeparator:
```
Comment (tarekgh) [inline src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs:112]: >UnicodeCategory.SpaceSeparator: [](http://example.com/codeflow?start=25&length=31)

what happen if having multiple spaces surrounded by letters? looks will be trimmed into one separator?

File: src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs [hunk 3]
```diff
@@ -21,149 +28,154 @@ public sealed override string ConvertName(string name)
...
+                {
+                    ExpandBuffer(ref destination);
+                }
+
+                destination[charsWritten++] = value;
+            }
+
+            void ExpandBuffer(ref Span<char> destination)
+            {
+                int newSize = checked(destination.Length * 2);
```
Comment (tarekgh) [inline src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs:155]: >int newSize = checked(destination.Length * 2); [](http://example.com/codeflow?start=16&length=46)

do we need to check the newSize if overflow occurred? or we expect ArrayPool<char>.Shared.Rent will throw when this happen?

**Review by tarekgh — APPROVED** [review]
Added a few questions, LGTM otherwise.

File: src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs (hunk 1 above)
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs:105]: Only letters can have casing.

File: src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs (hunk 3 above)
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs:155]: `checked` prevents overflows. The runtime will do it for you and throw an exception.

File: src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs:112]: Yes, this is by design. We have a few tests for that.

File: src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs:155]: I would guess passing a negative number to `ArrayPool.Rent` would throw some kind of ArgumentException, so it's preferable if we catch it here and throw OverflowException.

**Comment by eiriktsarpalis** [comment]
@YohDeadfall thank you for providing useful feedback!

**Pull request merged**